### PR TITLE
Enable Dual Indexable Object Header Shape Functionality

### PR DESCRIPTION
**Pull request by VermaSh**
Enable Dual Indexable Object Header Shape functionality introduced by
https://github.com/eclipse-openj9/openj9/pull/16096.

Revert changes in `jilconsts.c`:
  Constants in `jilconsts.c` were updated in previous PR:16096 to fix crypto
  crashes. These changes are no longer needed as the fix for crypto crash has
  been moved to internal repo.

Move `isIndexableDataAddrPresent` query to `j9ObjectModel`:
  Move `isIndexableDataAddrPresent` query to `j9ObjectModel`. This allows us to
  remove `ifdefs` in AOT and internal crypto evaluators.

Add new fields for array header shape and size to `vmInfo`:
  Add array header shape and size fields to `vmInfo` so that JIT Server can
  track different contiguous and discontiguous array header layouts.
  `isIndexableDataAddrPresent` is used to track if `dataAddr` field has been
  toggled. `_(dis)contiguousIndexableHeaderSize` are used to track size of
  contiguous and discontiguous array headers.

Update AOT to support dual indexable object header shape functionality.

Populate `dataAddr` field of multidimensional arrays on X and Z:
  Update `multianewArrayEvaluator` to populate `dataAddr` field of
  multidimensional arrays during inline array allocation. This change addresses
  following scenarios:
  - Both 1st and 2nd dimensional arrays are zero size.
  - 1st dimensional array is of size non-zero and 2nd dimensional array is
    of size zero.

Guard `dataAddr` initialization on AArch64, Z, X and Power:
  With dual header layout changes we can no longer assume `dataAddr` field
  to be present for all GC policies; we must call `isIndexableDataAddrPresent()`
  when inlining array allocation to check if the field has been toggled on or
  off.

Depends on https://github.com/eclipse-openj9/openj9/pull/16096/

### CODE REVIEW

**Comment by VermaSh** [comment]
This has all JIT changes needed to support dual indexable object header shape. I still need to add few other changes, for example, change to enable dual indexable object header shape by default. But JIT probably won't be needing anything else. So, when you get a chance please review these changes.

cc: @vijaysun-omr @jonoommen @0xdaryl @mpirvu @amicic

**Comment by VermaSh** [comment]
@vijaysun-omr @jonoommen @0xdaryl @mpirvu @amicic @zl-wang @r30shah @joransiu This is ready for review. I had to make minor changes to move couple of declarations into `ifdef` in J9TreeEvaluators on x and z since my last passing personal build build_info.php?build_id=48493. I have launched a new personal build to verify those changes build_info.php?build_id=48506

File: runtime/nls/jitm/j9jit.nls
Comment (mpirvu) [inline runtime/nls/jitm/j9jit.nls:394]: Missing endline at end of file.

**Review by mpirvu — COMMENTED** [review]
I didn't check the TreeEvaluators, but everything else looks good.

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4777]: If you are choosing to use multi-line statement, please follow how it is done at other places in the code base. You can find multiple example in code base  (e.g, https://github.com/eclipse-openj9/openj9/blob/master/runtime/compiler/optimizer/J9Inliner.cpp)

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4704]: Can we use conditional assignment here ?

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4772]: I would like to understand better purpose of this ASSERT. Is it the limitation on the instruction you are using that needs ASSERT of offset should always fit within 32-Bit?

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4784]: Can you put the statement in else in curly braces for symmetry?

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4785]: So in case of non 64-Bit, we would not generate this instruction, which looking at the original code was there, Is this intended? If yes, Can you put a small comment in the code or here in the PR to explain why ?

**Review by r30shah — CHANGES_REQUESTED** [review]

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4777]: I did it that way to make multi-line statement easier to read. Here's an instance where the benefit is more noticeable  https://github.com/eclipse-openj9/openj9/blob/ea4136ea95e21c50e441b50c3eaa891ca07a1c2f/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L10893-L10897

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4910]: Also now the unconditional branch to cFlowRegionEnd will only be generated for 64-Bit only., Is this intended?

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4941]: What are the typical limit for `contiguousArrayHeaderSizeInBytes` ? Can we use LA instead of LAY to save two bytes ?

**Review by zl-wang — APPROVED** [review]
LGTM

**Review by r30shah — CHANGES_REQUESTED** [review]

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4785]: ah sorry that is not intended, we want to generate the instruction in all cases when dataAddr pointer is not enabled. I'll correct the code.

**Review by r30shah — CHANGES_REQUESTED** [review]
@VermaSh I have checked out the Z Evaluator and made recommendations. Looking at the x86 evaluator briefly it looked similar some if not all suggestions applies there as well mostly related to code format and code that should be generated for 32 bit platform originally.

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4941]: Yup, we can. At the moment, max limit for `contiguousArrayHeaderSizeInBytes` is 24. So we should be ok using LA.

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4918]: I have fixed it in my latest commits

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4704]: Do you mean conditional assignment for `populateFirstDimDataAddrSlot`?

Comment (r30shah) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4704]: Yes

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4783]: Could I make a case to have the commas in front?
Having the comma at front would make it easier to differentiate between different assert arguments. It's similar to how having the conditional operators in front rather than at the end aids boolean expression readability.
```c++
TR_ASSERT_FATAL_WITH_NODE(node
      , ((TR::Compiler->om.compressObjectReferences()
            && (fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()) == 8)
        || (!TR::Compiler->om.compressObjectReferences()
            && fej9->getOffsetOfDiscontiguousDataAddrField() == fej9->getOffsetOfContiguousDataAddrField())
      , "Offset of dataAddr field in discontiguous array is expected to be 8 bytes more than contiguous array if using compressed refs, "
        "Or same if using full refs. "
        "But was %d bytes for discontigous and %d bytes for contiguous array.\n"
      , fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField());
```
```c++
TR_ASSERT_FATAL_WITH_NODE(node,
      ((TR::Compiler->om.compressObjectReferences()
            && (fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()) == 8)
      || (!TR::Compiler->om.compressObjectReferences()
            && fej9->getOffsetOfDiscontiguousDataAddrField() == fej9->getOffsetOfContiguousDataAddrField()),
      "Offset of dataAddr field in discontiguous array is expected to be 8 bytes more than contiguous array if using compressed refs, "
      "Or same if using full refs. "
      "But was %d bytes for discontigous and %d bytes for contiguous array.\n",
      fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField());
```

Comment (r30shah) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4783]: Hmm, I understand the point of making it readable but honestly, I really do not see big difference in the two version you posted (Though my view on Assert is that the condition we are checking before throwing assert should be readable in code which it is in both case). I am just being a nit-picky here, but the reason I made this suggestion was to have it similar to other places in code base.

**Comment by VermaSh** [comment]
Launched another personal build to verify new changes. `job/jvm.29.personal/32500/`

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (joransiu) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4782]: Minor improvement... If the offset difference is ever only going to be 0 or 8, we can use `LGHI` instead of `LGFI` and save 2-bytes in i-text.

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4768]: Just a question, is this the most appropriate location for the Assert (I believe it is not code-gen or this evaluator specific, right ? and have same FATAL assert in other code-gen as well)

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4776]: Should be `or`.

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4790]: We would not have a branch to `cFlowRegionEnd` generated when indexable data addr is set. can  we should be able to refactor this to avoid two `generateS390BranchInstruction` especially when both are unconditional branch ? Also, I would carefully check the Internal Control Flow region as well.

**Review by r30shah — CHANGES_REQUESTED** [review]

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4790]: > We would not have a branch to cFlowRegionEnd generated when indexable data addr is set.

That is intentional. `populateFirstDimDataAddrSlot` is right before `cFlowRegionEnd` so when dataAddr field is toggled on, we jump to `populateFirstDimDataAddrSlot` and then go to `cFlowRegionEnd` instead of jumping straight to `cFlowRegionEnd`.

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4768]: That's where we rely on the assumption about dataAddr field offset. I don't think we make that assumption anywhere else besides when inlining allocation of arrays (in `multianewArrayEvaluator`, `VMnewEvaluator` and couple of related helpers). So having the asserts there would make the broken code easy to find, but I can definitely move the asserts if there's a more appropriate location.

Comment (r30shah) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4768]: That is ok, I just asked

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4790]: Do we ever have instruction that branches to cFlowRegionEnd when indexable data addr is set?

Comment (VermaSh) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4790]: Nope, we always jump to `populateFirstDimDataAddrSlot` then go to `cFlowRegionEnd` when dataAddr field is toggled on.

Comment (r30shah) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4790]: Ok in that case, atleast what I can understand from the evaluator (we should verify this through the log file as well ), you do not need to generate `populateFirstDimDataAddrSlot` label symbol, and generate `cFlowRegionEnd` before you populate dataAddrSlot at line 4872. Based on your comment in https://github.com/eclipse-openj9/openj9/pull/16420/files#r1166123941, your ICF ends at `populateFirstDimDataAddrSlot` in case of `isIndexableDataAddrPresent` is set.

Again, this is simply based on the previous comment, I would like to take a look at the log file and verify this as well.

**Review by r30shah — CHANGES_REQUESTED** [review]

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (r30shah) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:4790]: Ok, in that case, you may even do not need to generate `populateFirstDimDataAddrSlot`. You should be able to generate `cFlowRegionEnd` before you populate the data Addr slot as that is where your ICF ends (This is purely based on what I can understand from the evaluator and your last comment, we should verify this through a log file for sanity as well).

**Review by r30shah — APPROVED** [review]
Z Codegen changes looks good to me.

**Comment by VermaSh** [comment]
I have updated `cFlowRegionEnd` and `fallThru` to `cFlowRegionDone` and `doneLabel`.

@0xdaryl / @vijaysun-omr can I please get a review?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (0xdaryl) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:6961]: If the minimum object size can change based on policy, keep in mind there are a few places in this file (and in the other backends) that use the configuration macro `J9_GC_MINIMUM_OBJECT_SIZE` (which is presently 16 on x86-64).  If you haven't already, you should find those places that use that macro and confirm that the logic is correct for dual header where the minimum size is 8.

**Review by 0xdaryl — COMMENTED** [review]
Are you planning on squashing the commits before this is merged?  Or at least reducing them into a more logical set of commits?

**Comment by VermaSh** [comment]
I had left `fixup!` commits to test out `git`'s `rebase --autosquash` feature so I don't have to squash manually after each review change. I have squashed the commits into a single commit. I'll launch another personal build for verification.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (VermaSh) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Sorry, there was a misunderstanding about minimum object size. `J9_GC_MINIMUM_OBJECT_SIZE` doesn't include `dataAddr` size so we'll never have an indexable object smaller than 16 bytes. I have updated my change to reflect that.

**Comment by VermaSh** [comment]
This is ready to for review/merge

**Comment by VermaSh** [comment]
@joransiu, @mpirvu Can I please get another review?

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by VermaSh** [comment]
Internal PRs (closedj9/pull/186 and closedj9/pull/194) must be merged before this to avoid crypto failures

**Review by joransiu — APPROVED** [review]

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk17

**Comment by 0xdaryl** [comment]
All failures appear to be due to UNB power outage affecting our Jenkins nodes.

**Comment by 0xdaryl** [comment]
Jenkins test sanity plinux,xlinux,amac,xmac jdk17

**Comment by 0xdaryl** [comment]
@VermaSh : please confirm it is safe to merge this PR now.

**Comment by VermaSh** [comment]
@0xdaryl Internal PR has been approved by everyone. Please go ahead with the merge. Internal PR will be safe to merge after this is merged.

File: runtime/oti/j9consts.h
Comment (keithc-ca) [inline runtime/oti/j9consts.h:485]: A C++ comment is not appropriate for a C header file. I'll create a PR to fix this, otherwise I expect we'll see compiles fail.

Comment (keithc-ca) [reply runtime/oti/j9consts.h:485]: See https://github.com/eclipse-openj9/openj9/pull/17342.

Comment (dmitripivkine) [reply runtime/oti/j9consts.h:485]: Technically this comment itself is not correct. J9_GC_MINIMUM_OBJECT_SIZE specifies minimum possible object size and has nothing to do with object headers details. I am suggesting to remove this comment at all

Comment (dmitripivkine) [reply runtime/oti/j9consts.h:485]: I can elaborate more:
- `J9_GC_MINIMUM_OBJECT_SIZE` just means GC would not expect end of the object (does not matter mixed or indexable) in address lower than `objectBeginningAddress + J9_GC_MINIMUM_OBJECT_SIZE`. There is only purpose for this definition

Comment (VermaSh) [reply runtime/oti/j9consts.h:485]: gotcha, creating a PR to fix it

Comment (dmitripivkine) [reply runtime/oti/j9consts.h:485]: please read comments above, PR has a been created already

**Pull request merged**
